File: analyze_cleanup.mjs
import fs from 'fs';
import readline from 'readline';

function normalize(s) {
  // The vivino names have literal \u0026 (backslash + u0026) stored as text
  // We need to replace that literal sequence with &
  let out = s;
  // Match literal backslash followed by u0026
  out = out.replace(/\x5cu0026/g, '&');
  out = out.replace(/&amp;/g, '&');
  return out.normalize('NFC').toLowerCase().trim();
}

async function main() {
  const rl = readline.createInterface({ input: fs.createReadStream('producer_winery_map.jsonl') });
  const slugMatches = [];
  const substringMatches = [];
  const suffixStripped = [];
  const allEntries = [];

  for await (const line of rl) {
    try {
      const j = JSON.parse(line);
      allEntries.push(j);
      if (j.match_confidence === 'slug_match') slugMatches.push(j);
      if (j.match_confidence === 'substring') substringMatches.push(j);
      if (j.match_confidence === 'suffix_stripped') suffixStripped.push(j);
    } catch {}
  }

  // Test normalization on first &-containing entry
  const testEntry = slugMatches.find(m => m.vivino_winery_name.includes('0026'));
  if (testEntry) {
    const rawVivino = testEntry.vivino_winery_name;
    console.log('=== NORMALIZATION TEST ===');
    console.log('Raw vivino name:', JSON.stringify(rawVivino));
    console.log('Char codes of first 20:', [...rawVivino].slice(0, 20).map(c => c.charCodeAt(0)).join(','));
    console.log('Normalized:', normalize(rawVivino));
    console.log('Producer normalized:', normalize(testEntry.producer_name));
    console.log('Match?', normalize(testEntry.producer_name) === normalize(testEntry.vivino_winery_name));
    console.log('');
  }

  // === SLUG_MATCH ANALYSIS ===
  const slugSafe = [];
  const slugFalsePositives = [];
  for (const m of slugMatches) {
    if (normalize(m.producer_name) === normalize(m.vivino_winery_name)) {
      slugSafe.push(m);
    } else {
      slugFalsePositives.push(m);
    }
  }

  console.log('=== SLUG_MATCH ANALYSIS ===');
  console.log(`Total: ${slugMatches.length}`);
  console.log(`Encoding-safe (names match after normalizing \\u0026→&): ${slugSafe.length}`);
  console.log(`FALSE POSITIVES (names genuinely different): ${slugFalsePositives.length}`);
  console.log('');
  if (slugFalsePositives.length > 0) {
    console.log('All false positive slug_matches:');
    for (const m of slugFalsePositives) {
      console.log(`  ✗ "${m.producer_name}" → "${normalize(m.vivino_winery_name)}" (winery ${m.vivino_winery_id}, ${m.wines_count} wines)`);
    }
  }

  // === SUFFIX_STRIPPED ANALYSIS ===
  console.log('');
  console.log('=== SUFFIX_STRIPPED ANALYSIS ===');
  console.log(`Total: ${suffixStripped.length}`);
  for (const m of suffixStripped) {
    console.log(`  "${m.producer_name}" → "${normalize(m.vivino_winery_name)}" (winery ${m.vivino_winery_id}, ${m.wines_count} wines)`);
  }

  // === SUBSTRING ANALYSIS ===
  const subSafe = [];
  const subRisky = [];
  for (const m of substringMatches) {
    const a = normalize(m.producer_name);
    const b = normalize(m.vivino_winery_name);
    // Safe if one fully contains the other
    if (a.includes(b) || b.includes(a)) {
      subSafe.push(m);
    } else {
      subRisky.push(m);
    }
  }

  console.log('');
  console.log('=== SUBSTRING ANALYSIS ===');
  console.log(`Total: ${substringMatches.length}`);
  console.log(`Contained (one name inside the other): ${subSafe.length}`);
  console.log(`Risky (partial/weak overlap): ${subRisky.length}`);
  console.log('');
  if (subRisky.length > 0) {
    console.log('Risky substring matches:');
    for (const m of subRisky) {
      console.log(`  ? "${m.producer_name}" → "${normalize(m.vivino_winery_name)}" (winery ${m.vivino_winery_id}, ${m.wines_count} wines)`);
    }
  }

  // === SUMMARY ===
  const totalResolved = allEntries.filter(e => e.vivino_winery_id).length;
  const totalNone = allEntries.filter(e => e.match_confidence === 'none').length;
  const totalExact = allEntries.filter(e => e.match_confidence === 'exact').length;

  console.log('');
  console.log('=== SUMMARY ===');
  console.log(`Total producers: ${allEntries.length}`);
  console.log(`Exact matches (safe): ${totalExact}`);
  console.log(`Slug_match encoding-safe: ${slugSafe.length}`);
  console.log(`Slug_match FALSE POSITIVES: ${slugFalsePositives.length}`);
  console.log(`Suffix_stripped: ${suffixStripped.length}`);
  console.log(`Substring contained (likely safe): ${subSafe.length}`);
  console.log(`Substring risky: ${subRisky.length}`);
  console.log(`Unresolved: ${totalNone}`);
  console.log('');
  console.log(`ENTRIES TO DELETE: ${slugFalsePositives.length} slug_match FPs + ${subRisky.length} risky substring = ${slugFalsePositives.length + subRisky.length}`);
  console.log(`ENTRIES TO KEEP: ${totalExact} exact + ${slugSafe.length} slug safe + ${subSafe.length} substring safe + ${suffixStripped.length} suffix = ${totalExact + slugSafe.length + subSafe.length + suffixStripped.length}`);
}

main();

File: scripts_archive/vivino/analyze_cleanup.py
"""
Analyzes slug_match/substring/suffix_stripped match quality from JSONL.

Reads producer_winery_map.jsonl and categorizes matches by quality:
- slug_match: safe (encoding) vs false positives (genuinely different names)
- substring: contained (safe) vs risky (partial overlap)
- suffix_stripped: all entries

Usage:
    python -m pipeline.vivino.analyze_cleanup
"""

import sys
import json
import unicodedata
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parents[2]))


def normalize(s):
    """Normalize handling literal \\u0026 sequences and &amp;"""
    out = s.replace("\\u0026", "&").replace("&amp;", "&")
    return unicodedata.normalize("NFC", out).lower().strip()


def main():
    slug_matches = []
    substring_matches = []
    suffix_stripped = []
    all_entries = []

    with open("producer_winery_map.jsonl", "r", encoding="utf-8") as f:
        for line in f:
            line = line.strip()
            if not line:
                continue
            try:
                j = json.loads(line)
            except json.JSONDecodeError:
                continue
            all_entries.append(j)
            mc = j.get("match_confidence")
            if mc == "slug_match":
                slug_matches.append(j)
            elif mc == "substring":
                substring_matches.append(j)
            elif mc == "suffix_stripped":
                suffix_stripped.append(j)

    # Normalization test
    test_entry = next((m for m in slug_matches if "0026" in m.get("vivino_winery_name", "")), None)
    if test_entry:
        print("=== NORMALIZATION TEST ===")
        print(f"Raw vivino name: {json.dumps(test_entry['vivino_winery_name'])}")
        print(f"Normalized: {normalize(test_entry['vivino_winery_name'])}")
        print(f"Producer normalized: {normalize(test_entry['producer_name'])}")
        print(f"Match? {normalize(test_entry['producer_name']) == normalize(test_entry['vivino_winery_name'])}")
        print()

    # SLUG_MATCH ANALYSIS
    slug_safe = []
    slug_false_positives = []
    for m in slug_matches:
        if normalize(m["producer_name"]) == normalize(m["vivino_winery_name"]):
            slug_safe.append(m)
        else:
            slug_false_positives.append(m)

    print("=== SLUG_MATCH ANALYSIS ===")
    print(f"Total: {len(slug_matches)}")
    print(f"Encoding-safe (names match after normalizing): {len(slug_safe)}")
    print(f"FALSE POSITIVES (names genuinely different): {len(slug_false_positives)}")
    print()
    if slug_false_positives:
        print("All false positive slug_matches:")
        for m in slug_false_positives:
            print(f'  x "{m["producer_name"]}" -> "{normalize(m["vivino_winery_name"])}" (winery {m["vivino_winery_id"]}, {m.get("wines_count", 0)} wines)')

    # SUFFIX_STRIPPED ANALYSIS
    print()
    print("=== SUFFIX_STRIPPED ANALYSIS ===")
    print(f"Total: {len(suffix_stripped)}")
    for m in suffix_stripped:
        print(f'  "{m["producer_name"]}" -> "{normalize(m["vivino_winery_name"])}" (winery {m["vivino_winery_id"]}, {m.get("wines_count", 0)} wines)')

    # SUBSTRING ANALYSIS
    sub_safe = []
    sub_risky = []
    for m in substring_matches:
        a = normalize(m["producer_name"])
        b = normalize(m["vivino_winery_name"])
        if a in b or b in a:
            sub_safe.append(m)
        else:
            sub_risky.append(m)

    print()
    print("=== SUBSTRING ANALYSIS ===")
    print(f"Total: {len(substring_matches)}")
    print(f"Contained (one name inside the other): {len(sub_safe)}")
    print(f"Risky (partial/weak overlap): {len(sub_risky)}")
    print()
    if sub_risky:
        print("Risky substring matches:")
        for m in sub_risky:
            print(f'  ? "{m["producer_name"]}" -> "{normalize(m["vivino_winery_name"])}" (winery {m["vivino_winery_id"]}, {m.get("wines_count", 0)} wines)')

    # SUMMARY
    total_resolved = sum(1 for e in all_entries if e.get("vivino_winery_id"))
    total_none = sum(1 for e in all_entries if e.get("match_confidence") == "none")
    total_exact = sum(1 for e in all_entries if e.get("match_confidence") == "exact")

    print()
    print("=== SUMMARY ===")
    print(f"Total producers: {len(all_entries)}")
    print(f"Exact matches (safe): {total_exact}")
    print(f"Slug_match encoding-safe: {len(slug_safe)}")
    print(f"Slug_match FALSE POSITIVES: {len(slug_false_positives)}")
    print(f"Suffix_stripped: {len(suffix_stripped)}")
    print(f"Substring contained (likely safe): {len(sub_safe)}")
    print(f"Substring risky: {len(sub_risky)}")
    print(f"Unresolved: {total_none}")
    print()
    delete_count = len(slug_false_positives) + len(sub_risky)
    keep_count = total_exact + len(slug_safe) + len(sub_safe) + len(suffix_stripped)
    print(f"ENTRIES TO DELETE: {len(slug_false_positives)} slug_match FPs + {len(sub_risky)} risky substring = {delete_count}")
    print(f"ENTRIES TO KEEP: {total_exact} exact + {len(slug_safe)} slug safe + {len(sub_safe)} substring safe + {len(suffix_stripped)} suffix = {keep_count}")


if __name__ == "__main__":
    main()
